src/__tests__/e2e/lsp-client.ts

```diff
@@ -0,0 +1,378 @@
+    return new Promise((resolve, reject) => {
+      this.server = spawn("bun", ["run", "src/server/cli.ts", "--stdio"], {
+        stdio: ["pipe", "pipe", "pipe"],
+        env: { ...process.env, NODE_ENV: "test" },
+      });
+
+      if (!this.server.stdin || !this.server.stdout || !this.server.stderr) {
+        reject(new Error("Failed to create server process streams"));
+        return;
+      }
+
+      this.server.stdout.on("data", (data: Buffer) => {
+        this.handleServerMessage(data.toString());
+      });
+
+      this.server.stderr.on("data", (data: Buffer) => {
+        console.error("Server stderr:", data.toString());
+      });
+
+      this.server.on("error", (error) => {
+        reject(error);
+      });
+
+      this.server.on("exit", (code) => {
+        // console.log(`Server exited with code ${code}`);
+      });
+
+      // Give server time to start
+      setTimeout(resolve, 100);
+    });
+  }
+
+  async stop(): Promise<void> {
+    if (this.server) {
+      this.server.kill();
+      this.server = null;
+    }
+    this.pendingRequests.clear();
+    this.serverMessages = [];
+    this.numRefreshRequest = 0;
+    this.currentSettings = {};
+  }
+
+  private handleServerMessage(data: string): void {
+    this.buffer += data;
+
+    // Process complete messages
+    while (true) {
+      const headerEnd = this.buffer.indexOf("\r\n\r\n");
+      if (headerEnd === -1) break;
+
+      const headerPart = this.buffer.substring(0, headerEnd);
+      const contentLengthMatch = headerPart.match(/Content-Length: (\d+)/);
+
+      if (!contentLengthMatch) {
+        console.error("Invalid LSP message format");
+        this.buffer = this.buffer.substring(headerEnd + 4);
+        continue;
+      }
+
+      const contentLength = parseInt(contentLengthMatch[1] ?? "0");
+      const messageStart = headerEnd + 4;
+
+      if (this.buffer.length < messageStart + contentLength) {
+        // Wait for complete message
+        break;
+      }
+
+      const messageContent = this.buffer.substring(messageStart, messageStart + contentLength);
+      this.buffer = this.buffer.substring(messageStart + contentLength);
+
+      try {
+        const message: LSPMessage = JSON.parse(messageContent);
+        this.serverMessages.push(message);
+        this.handleMessage(message);
+      } catch (error) {
+        console.error("Failed to parse LSP message:", error, "Content:", messageContent);
+      }
+    }
+  }
+
+  private handleMessage(message: LSPMessage): void {
+    if (message.id && this.pendingRequests.has(message.id as number)) {
+      const request = this.pendingRequests.get(message.id as number)!;
+      this.pendingRequests.delete(message.id as number);
+
+      if (message.error) {
+        request.reject(new Error(`LSP Error: ${message.error.message}`));
+      } else {
+        request.resolve(message.result);
+      }
+    } else if (message.method) {
+      if (message.method === 'workspace/diagnostic/refresh') {
```

Comment:
IDK about you, I'd prefer if these strings were replaced with imports from the protocol. For example, this string is `DiagnosticRefreshRequest.method`

Reply:
That's fine with me!